### ISSUE TITLE
Tensors with multiple unbacked SymInt dims don't work in Inductor

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#120548]
### 🐛 Describe the bug

Internal xref: https://fb.workplace.com/groups/6829516587176185/posts/6892427944218382/

The use case here is modeling 2D/3D variable size data (e.g., images/videos of varying input dimensionality)

This doesn't work:

```
import torch
import torch._dynamo.config

torch._dynamo.config.capture_scalar_outputs = True

@torch.compile(fullgraph=True)
def f(x):
    s = x.tolist()
    return torch.empty(*s)

f(torch.tensor([3, 4, 5]))
```

Fails with

```
  File "/data/users/ezyang/a/pytorch/torch/_inductor/graph.py", line 976, in run_node
    dense = torch._prims_common.is_non_overlapping_and_dense(n.meta["val"])
  File "/data/users/ezyang/a/pytorch/torch/_prims_common/__init__.py", line 356, in is_non_overlapping_and_dense
    if is_contiguous(a) or is_channels_last_contiguous(a):
  File "/data/users/ezyang/a/pytorch/torch/_prims_common/__init__.py", line 244, in is_contiguous
    if y != expected_stride:
  File "/data/users/ezyang/a/pytorch/torch/__init__.py", line 374, in __bool__
    return self.node.bool_()
  File "/data/users/ezyang/a/pytorch/torch/fx/experimental/sym_node.py", line 432, in bool_
    return self.guard_bool("", 0)
  File "/data/users/ezyang/a/pytorch/torch/fx/experimental/sym_node.py", line 374, in guard_bool
    r = self.shape_env.evaluate_expr(self.expr, self.hint, fx_node=self.fx_node)
  File "/data/users/ezyang/a/pytorch/torch/fx/experimental/recording.py", line 255, in wrapper
    return event.run(self)
  File "/data/users/ezyang/a/pytorch/torch/fx/experimental/recording.py", line 156, in run
    return self.f(*args, **kwargs)
  File "/data/users/ezyang/a/pytorch/torch/fx/experimental/symbolic_shapes.py", line 3914, in evaluate_expr
    raise self._make_data_dependent_error(expr.xreplace(self.var_to_val), expr)
torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
GuardOnDataDependentSymNode: Could not guard on data-dependent expression Ne(Max(1, u2), u2) (unhinted: Ne(Max(1, u2), u2)).  (Size-like symbols: u2)

Potential framework code culprit (scroll up for full backtrace):
  File "/data/users/ezyang/a/pytorch/torch/_prims_common/__init__.py", line 244, in is_contiguous
    if y != expected_stride:

For more information, run with TORCH_LOGS="dynamic"
For extended logs when we create symbols, also add TORCHDYNAMO_EXTENDED_DEBUG_CREATE_SYMBOL="u2"
If you suspect the guard was triggered from C++, add TORCHDYNAMO_EXTENDED_DEBUG_CPP=1
For more debugging help, see https://docs.google.com/document/d/1HSuTTVvYH1pTew89Rtpeu84Ht3nQEFTYhAX3Ypa_xJs/edit?usp=sharing
```

Notably, graph capture is all fine; we are only choking in Inductor. cc @gchanan @zou3519 @kadeng @msaroufim @bdhirsh @anijain2305 @chauhang @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire

It is tempting to add guard_size_oblivious to torch/_prims_common/__init__.py but this is a bit questionable: the tensors we create here should be obviously contiguous (since we built them with torch.empty(*s)), and adding that here will cause us to report that it is NOT contiguous, which is not what we want. Here's a better fix:

```
diff --git a/torch/_inductor/graph.py b/torch/_inductor/graph.py
index 51af2a51c49..fb8810a94e3 100644
--- a/torch/_inductor/graph.py
+++ b/torch/_inductor/graph.py
@@ -973,7 +973,7 @@ class GraphLowering(torch.fx.Interpreter):
                 n.meta["val"], torch.Tensor
             ):
                 strides = n.meta["val"].stride()
-                dense = torch._prims_common.is_non_overlapping_and_dense(n.meta["val"])
+                dense = torch.ops.aten.is_non_overlapping_and_dense.default(n.meta["val"])
                 # requiring a stride order for a non-dense output wouldn't
                 # recreate the same strides, and would fail with view, defer for now.
                 if dense and len(strides):
```

We then choke at

```
  File "/data/users/ezyang/a/pytorch/torch/_inductor/graph.py", line 980, in run_node                                                        
    stride_order = ir.get_stride_order(strides)                                                                                              
  File "/data/users/ezyang/a/pytorch/torch/_inductor/ir.py", line 198, in get_stride_order                                                   
    sorted_idx: List[int] = argsort(seq)                                                                                                     
  File "/data/users/ezyang/a/pytorch/torch/_inductor/utils.py", line 762, in argsort                                                         
    return list(reversed(sorted(a_r, key=getter, reverse=True)))  # noqa: C413                                                               
  File "/data/users/ezyang/a/pytorch/torch/__init__.py", line 374, in __bool__           
    return self.node.bool_()  
  File "/data/users/ezyang/a/pytorch/torch/fx/experimental/sym_node.py", line 432, in bool_  
    return self.guard_bool("", 0)
  File "/data/users/ezyang/a/pytorch/torch/fx/experimental/sym_node.py", line 374, in guard_bool
    r = self.shape_env.evaluate_expr(self.expr, self.hint, fx_node=self.fx_node)                                                             
  File "/data/users/ezyang/a/pytorch/torch/fx/experimental/recording.py", line 255, in wrapper            
    return event.run(self)                                            
  File "/data/users/ezyang/a/pytorch/torch/fx/experimental/recording.py", line 156, in run
    return self.f(*args, **kwargs)
  File "/data/users/ezyang/a/pytorch/torch/fx/experimental/symbolic_shapes.py", line 3914, in evaluate_expr        
    raise self._make_data_dependent_error(expr.xreplace(self.var_to_val), expr)            
torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:                                                                          
GuardOnDataDependentSymNode: Could not guard on data-dependent expression Max(1, u1)*Max(1, u2) < Max(1, u2) (unhinted: Max(1, u1)*Max(1, u2)
 < Max(1, u2)).  (Size-like symbols: u1, u2) 
```

So it seems we need to thread the needle and avoid attempting to directly compute ordering on strides if we actually know it's contiguous out of band.

Also cc @shunting314 for NHWC striding optimization.

cc @msaroufim @bdhirsh @anijain2305 @zou3519 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames 

### Versions

main

**Comment by ezyang** [comment]
Yes. Actually, you kind of already know because you can safely query is non-overlapping and dense on the meta tensor (as long as you don't use the prim_util--that's a footgun we need to do something about). It's just about writing the code carefully.

**Comment by ezyang** [comment]
This problem also affected @kimishpatel in llama export

**Comment by shunting314** [comment]
@ezyang  do you know where is the torch.ops.aten.is_non_overlapping_and_dense.default op implementation for FakeTensor?

BTW, even if we don't do any channels last optimization, we will still run into this issue since inductor need know stride order to make sure we don't change user-visible tensor's strides.

**Comment by ezyang** [comment]
It's native in C++, it just asks TensorImpl what it thinks, and that just reads out the value from a cached field.

**Comment by ColinPeppler** [comment]
Took a shot at skipping the stride order if we know the meta tensor is contiguous. https://github.com/pytorch/pytorch/pull/120688

But I run into 
```
$ python test/inductor/test_unbacked_symints.py -k test_expand_cuda

  File "/data/users/colinpeppler/pytorch/torch/_inductor/graph.py", line 976, in run_node
    dense = torch.ops.aten.is_non_overlapping_and_dense(n.meta["val"])
  File "/data/users/colinpeppler/pytorch/torch/_ops.py", line 825, in __call__
    return self_._op(*args, **(kwargs or {}))
  File "/data/users/colinpeppler/pytorch/torch/fx/experimental/sym_node.py", line 374, in guard_bool
    r = self.shape_env.evaluate_expr(self.expr, self.hint, fx_node=self.fx_node)
  File "/data/users/colinpeppler/pytorch/torch/fx/experimental/recording.py", line 255, in wrapper
    return event.run(self)
  File "/data/users/colinpeppler/pytorch/torch/fx/experimental/recording.py", line 156, in run
    return self.f(*args, **kwargs)
  File "/data/users/colinpeppler/pytorch/torch/fx/experimental/symbolic_shapes.py", line 3914, in evaluate_expr
    raise self._make_data_dependent_error(expr.xreplace(self.var_to_val), expr)
torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
GuardOnDataDependentSymNode: Could not guard on data-dependent expression Eq(IsNonOverlappingAndDenseIndicator(u0, 128, 1, 0), 1) (unhinted: Eq(IsNonOverlappingAndDenseIndicator(u0, 128, 1, 0), 1)).  (Size-like symbols: u0)
```

`torch.ops.aten.is_non_overlapping_and_dense` tries to evaluate `Eq(IsNonOverlappingAndDenseIndicator(u0, 128, 1, 0), 1)` but fails. I'm guessing it's because we have [here](https://github.com/pytorch/pytorch/blob/be31e522ce33f8fc5d53dd0681e9e0f796dd7a90/c10/core/TensorImpl.h#L866-L869).

**Comment by ezyang** [comment]
I guess, for this case, if we see that any stride is zero for a non-zero size dim, then we are definitely non-contiguous. But per discussion with @avikchaudhuri it might just be better to force some simplified guards when you actually end up triggering non-overlapping and dense.

**Comment by ezyang** [comment]
Here is an example of a branchy implementation

```
template <typename T>
inline bool is_channels_last_strides_2d_s4(
    const ArrayRef<T> sizes,
    const ArrayRef<T> strides) {
  T min = 0;
  // special case for trivial C dimension. default to NCHW
  if (strides[1] == 0) {
    return false;
  }
  // loop strides indices
  for (auto& d : {1, 3, 2, 0}) {
    if (sizes[d] == 0) {
      return false;
    }
    if (strides[d] < min) {
      return false;
    }
    // Fallback to NCHW as default layout for ambiguous cases
    // This is the flaw of implicit memory_format from strides.
    // N111 tensor with identical strides for size 1 dimension;
    // Two cases could lead us here:
    // a. N111 contiguous Tensor ([N,1,1,1]@[1,1,1,1])
    // b. N11W contiguous Tensor sliced on the W-dimension.
    // ([N,1,1,1]@[W,W,W,W])
    if (d == 0 && min == strides[1]) {
      return false;
    }
    // This is necessary to:
    // 1. distinguish the memory_format of N1H1;
    //     [H, 1, 1, 1] channels_last stride
    //     [H, H, 1, 1] contiguous stride
    // 2. permutation of 1C1W:
    //     [1, C, 1, H]@[HC, H, H, 1] transpose(1, 3)
    //     [1, H, 1, C]@[HC, 1, H, H] shouldn't be identified as channels_last
    min = strides[d];
    if (sizes[d] > 1) {
      min *= sizes[d];
    }
  }
  return true;
}
```

and a branch free implementation

```
def sympy_is_channels_last_strides_generic(sizes, strides, dim_order):
    import sympy

    dim = len(sizes)

    if dim != len(dim_order):
        return sympy.false

    m = sympy.Integer(0)
    r = sympy.true

    # special case for trivial C dimension. default to NCHW
    r &= sympy.Ne(strides[1], 0)

    for d in dim_order:
        r &= sympy.Ne(sizes[d], 0) & (strides[d] >= m)
        # Fallback to NCHW as default layout for ambiguous cases
        # This is the flaw of implicit memory_format from strides.
        # N111 tensor with identical strides for size 1 dimension;
        # Two cases could lead us here:
        # a. N111 contiguous Tensor ([N,1,1,1]@[1,1,1,1])
        # b. N11W contiguous Tensor sliced on the W-dimension.
        # ([N,1,1,1]@[W,W,W,W])
        if d == 0:
            r &= sympy.Ne(m, strides[1])
        # This is necessary to:
        # 1. distinguish the memory_format of N1H1;
        #     [H, 1, 1, 1] channels_last stride
        #     [H, H, 1, 1] contiguous stride
        # 2. permutation of 1C1W:
        #     [1, C, 1, H]@[HC, H, H, 1] transpose(1, 3)
        #     [1, H, 1, C]@[HC, 1, H, H] shouldn't be identified as
        #     channels_last
        m = strides[d] * sympy.Max(sizes[d], 1)

    return r
```

Notice that the branch free implementation doesn't have any conditionals on sizes, while the branchy implementation does. The branch-free implementation doesn't impose any guards, but you now end up with a complicated expression with conjunctions and disjunctions that you may eventually have to evaluate the truth of. This expression can be difficult for people to understand.

Whereas in the branchy implementation, you'd get a number of guards like `sizes[d] > 1`, all individually simple, and then eventually you'd get True/False. This is much easier to interpret if you're `torch._check` ing your way to victory.

**Comment by ezyang** [comment]
@ColinPeppler maybe we should land your partial fix anyway, even if it doesn't fix the whole thing

**Comment by ColinPeppler** [comment]
having a WIP PR that isn't the cleanest, but would like some thoughts

https://github.com/pytorch/pytorch/pull/122369

**Comment by ColinPeppler** [comment]
closing. fixed by https://github.com/pytorch/pytorch/issues/122296

**Comment by ColinPeppler** [comment]
will look into using `torch.ops.aten.is_non_overlapping_and_dense` over `torch._prims_common.is_non_overlapping_and_dense` at some point
